### PR TITLE
Pseudorandom dataframe split

### LINKED ISSUE #252: Support for randomized partitions of datasets
I'd love to see future capability that allows one to perform a randomized partition of a dataset and follow-on computation on the partitions. This is critical for simple hold-out tests and cross-validation to assess the generalization performance of models resulting from machine learning.

### PR DESCRIPTION
**Pull request by mrocklin**
Fixes #252 

This adds a `random_split` method to `dask.dataframe.DataFrame` that partitions a dataframe into multiple disjoint parts.  One can specify a preferred split (e.g. 80/20 or 33/33/34) as a list of fractions.

Example
------

```python
In [1]: import dask.dataframe as dd

In [2]: df = dd.read_csv('trip_data_1_*.csv')

In [3]: a, b = df.random_split([0.9, 0.1])  # 90/10 split

In [4]: a.head(3)
Out[4]: 
                          medallion                      hack_license  \
0  89D227B655E5C82AECF13C3F540D4CF4  BA96DE419E711691B9445D6A6307C170   
1  0BD7C8F5BA12B88E0B67BED28BEA73D8  9FD8F69F0804BDB5549F40E9DA1BE472   
2  0BD7C8F5BA12B88E0B67BED28BEA73D8  9FD8F69F0804BDB5549F40E9DA1BE472   

  vendor_id  rate_code store_and_fwd_flag      pickup_datetime  \
0       CMT          1                  N  2013-01-01 15:11:48   
1       CMT          1                  N  2013-01-06 00:18:35   
2       CMT          1                  N  2013-01-05 18:49:41   

      dropoff_datetime  passenger_count  trip_time_in_secs  trip_distance  \
0  2013-01-01 15:18:10                4                382            1.0   
1  2013-01-06 00:22:54                1                259            1.5   
2  2013-01-05 18:54:23                1                282            1.1   

   pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  
0        -73.978165        40.757977         -73.989838         40.751171  
1        -74.006683        40.731781         -73.994499         40.750660  
2        -74.004707        40.737770         -74.009834         40.726002  

In [5]: b.head(3)
Out[5]: 
                           medallion                      hack_license  \
5   20D9ECB2CA0767CF7A01564DF2844A3E  598CCE5B9C1918568DEE71F43CF26CD2   
34  7E3256C342CAFB3C23D3ABB98B9F3FB6  4F7873C913735088F9BEF85C1D054D31   
59  82420BD631D2BB88E2F213A8CA15BECE  213779F7AA40E58009B3457E3FA7089D   

   vendor_id  rate_code store_and_fwd_flag      pickup_datetime  \
5        CMT          1                  N  2013-01-07 15:27:48   
34       CMT          1                  N  2013-01-05 03:00:46   
59       CMT          1                  N  2013-01-08 10:38:22   

       dropoff_datetime  passenger_count  trip_time_in_secs  trip_distance  \
5   2013-01-07 15:38:37                1                648            1.7   
34  2013-01-05 03:23:28                1               1361            5.5   
59  2013-01-08 10:56:19                1               1076            3.1   

    pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  
5         -73.966743        40.764252         -73.983322         40.743763  
34        -73.965500        40.711113         -73.918327         40.758671  
59        -73.967682        40.751221         -73.995232         40.717422  
```

The resulting split dataframes share the same graph so, if you compute them together you'll benefit from single pass computation (depending on the scheduler handling this well.)

cc @diehl

cc @jreback you might find the pd_split function of use

### CODE REVIEW

**Comment by jreback** [comment]
@mrocklin see here - http://pandas.pydata.org/pandas-docs/stable/whatsnew.html#whatsnew-0161-enhancements-sample

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -805,4 +831,28 @@ def get(dsk, keys, get=get_sync, **kwargs):
...
+
+    >>> b
+       a  b
+    0  1  2
+    3  4  5
+    4  5  6
+    """
+    p = list(p)
+    index = pseudorandom(len(df), p, seed)
+    return [df.iloc[index == i] for i in range(len(p))]
```
Comment (jreback) [inline dask/dataframe/core.py]: use ``df.take(list(range(len(p)))``, must faster

**Comment by shoyer** [comment]
@jreback This is similar but not quite the same as `sample`... it could be a nice complement  for pandas. Though scikit-learn's `train_test_split` largely solves the same problem.

File: dask/dataframe/core.py (hunk 1 above)
Comment (cpcloud) [inline dask/dataframe/core.py]: that doesn't do the same thing. this function returns a list of dataframes, whereas `df.take` with a list returns a single dataframe. i assume `len(p)` is generally going to be small, i.e., the number of splits will be small.

Comment (jreback) [inline dask/dataframe/core.py]: ok sorry

``[ df.take(i) for i in range(len(p)) ]``. Using a boolean selection and iloc is pretty slow compared to ``.take``

**Comment by jreback** [comment]
@shoyer I know, we really should just add ``.split/.partition`` already

File: dask/dataframe/core.py (hunk 1 above)
Comment (cpcloud) [inline dask/dataframe/core.py]: that's still not the same thing. here's an example:

let's say that `p = [0.5, 0.5]`

```
index = np.array([0, 0, 1, 0, 1, 0, 0, 1, 1, 1])
df = pd.DataFrame(np.random.randn(10, 2), columns=list('ab'))
[df.take(0), df.take(1)] != [df.iloc[index == 0], df.iloc[index == 1]]
```

Comment (jreback) [inline dask/dataframe/core.py]: right, the ``iloc[...]`` does translate to the ``.take``, nvm.

**Comment by mrocklin** [comment]
@diehl any thoughts?  Is this the right user interface?  Does it do what you expect?

**Comment by diehl** [comment]
@mrocklin The .random_split method looks great. That is exactly the functionality I'd need to take an overall dataset and split it into three separate dataframes for training, validation and test data.
